In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('train.csv')
df.head()

filename  opacity  diabetic retinopathy  glaucoma  macular edema  \
0  c24a1b14d253.jpg        0                     0         0              0   
1  9ee905a41651.jpg        0                     0         0              0   
2  3f58d128caf6.jpg        0                     0         1              0   
3  4ce6599e7b20.jpg        1                     0         0              0   
4  0def470360e4.jpg        1                     0         0              0   

   macular degeneration  retinal vascular occlusion  normal  
0                     0                           1       0  
1                     0                           1       0  
2                     0                           0       0  
3                     1                           0       0  
4                     1                           0       0

In [3]:
df.shape

(3435, 8)

In [4]:
df.columns[1:]

Index(['opacity', 'diabetic retinopathy', 'glaucoma', 'macular edema',
       'macular degeneration', 'retinal vascular occlusion', 'normal'],
      dtype='object')

In [5]:
df.dtypes

filename                      object
opacity                        int64
diabetic retinopathy           int64
glaucoma                       int64
macular edema                  int64
macular degeneration           int64
retinal vascular occlusion     int64
normal                         int64
dtype: object

In [6]:
for i in df.columns[1:]:
    print(i, df[i].sum())

opacity 1533
diabetic retinopathy 755
glaucoma 597
macular edema 516
macular degeneration 574
retinal vascular occlusion 440
normal 525


### Single Occurences

In [7]:
df[(df.sum(axis=1) == 1) & (df.normal == 0)].shape

(1706, 8)

In [8]:
df[df.sum(axis=1)==4].shape

(13, 8)

### Number of Cases with Multiple Occurences of Abnormalities in Images
- there are 3435 Cases
- 2231 Cases with only 1  occurences, 1706 Abnormal and 525 healthy
- 916 Cases with 2 Abnormalities
- 275 Cases with 3 Abnormalities
- 13 Cases with 4 Abnormalities

### Abnormalities that occur together

In [9]:
df['new'] = df.eq(1).apply(lambda x: ', '.join(x[x].index), axis = 1)

In [10]:
df.head()

filename  opacity  diabetic retinopathy  glaucoma  macular edema  \
0  c24a1b14d253.jpg        0                     0         0              0   
1  9ee905a41651.jpg        0                     0         0              0   
2  3f58d128caf6.jpg        0                     0         1              0   
3  4ce6599e7b20.jpg        1                     0         0              0   
4  0def470360e4.jpg        1                     0         0              0   

   macular degeneration  retinal vascular occlusion  normal  \
0                     0                           1       0   
1                     0                           1       0   
2                     0                           0       0   
3                     1                           0       0   
4                     1                           0       0   

                             new  
0     retinal vascular occlusion  
1     retinal vascular occlusion  
2                       glaucoma  
3  opacity, macular degeneration  
4  opacity, macular degeneration

In [11]:
df.new.value_counts()

opacity                                                                     671
normal                                                                      525
glaucoma                                                                    327
macular degeneration                                                        299
diabetic retinopathy                                                        212
opacity, macular degeneration                                               212
opacity, diabetic retinopathy                                               186
retinal vascular occlusion                                                  170
diabetic retinopathy, macular edema                                         162
opacity, diabetic retinopathy, macular edema                                154
opacity, glaucoma                                                           140
macular edema, retinal vascular occlusion                                    74
opacity, macular edema, retinal vascular

In [12]:
df[(df.sum(axis=1)==3) | (df.sum(axis=1)==4)].new.value_counts()

opacity, diabetic retinopathy, macular edema                                154
opacity, macular edema, retinal vascular occlusion                           62
opacity, glaucoma, macular degeneration                                      16
glaucoma, macular edema, retinal vascular occlusion                          14
opacity, glaucoma, retinal vascular occlusion                                10
opacity, glaucoma, macular edema, retinal vascular occlusion                  8
opacity, diabetic retinopathy, macular degeneration                           5
opacity, diabetic retinopathy, retinal vascular occlusion                     4
diabetic retinopathy, glaucoma, macular edema                                 3
opacity, macular degeneration, retinal vascular occlusion                     3
glaucoma, macular degeneration, retinal vascular occlusion                    2
opacity, diabetic retinopathy, macular edema, macular degeneration            2
opacity, diabetic retinopathy, macular e

### Augmentation: Class Imbalance between Classes: Idea 1: Eliminate Instances with 3 or 4 labels

In [13]:
df_small = df[(df.sum(axis=1)==2) | (df.sum(axis=1)==1)]

In [14]:
df_small.new.value_counts()

opacity                                             671
normal                                              525
glaucoma                                            327
macular degeneration                                299
diabetic retinopathy                                212
opacity, macular degeneration                       212
opacity, diabetic retinopathy                       186
retinal vascular occlusion                          170
diabetic retinopathy, macular edema                 162
opacity, glaucoma                                   140
macular edema, retinal vascular occlusion            74
opacity, retinal vascular occlusion                  55
glaucoma, macular degeneration                       31
glaucoma, retinal vascular occlusion                 27
macular edema                                        27
diabetic retinopathy, glaucoma                       15
diabetic retinopathy, retinal vascular occlusion      5
glaucoma, macular edema                         

In [19]:
df_small.shape

(3147, 9)

### For EDA: new cols for combinations

In [15]:
# iterate through rows, for 1 in col add colnames to new col and 1 as value

# maybe not

### Pearsons Correlation between Abnormalities

In [16]:
df.corr().abs().unstack().sort_values(ascending=False).drop_duplicates()

opacity                     opacity                       1.000000
diabetic retinopathy        macular edema                 0.418289
normal                      opacity                       0.381328
macular edema               retinal vascular occlusion    0.233828
diabetic retinopathy        normal                        0.225444
                            macular degeneration          0.220781
glaucoma                    diabetic retinopathy          0.208188
normal                      glaucoma                      0.194811
                            macular degeneration          0.190252
macular degeneration        macular edema                 0.183955
normal                      macular edema                 0.178583
diabetic retinopathy        retinal vascular occlusion    0.176092
retinal vascular occlusion  normal                        0.162803
macular degeneration        retinal vascular occlusion    0.155337
glaucoma                    opacity                       0.14

Obwohl opacity und makular degeneration am häufigsten zusammen auftreten, haben sie eine geringe Corr. ist das seltsam? wie wird corr zwishen bools berechnet mit pearson?

### Save Datasets
- df is dataset with new column containing all labels
- df_small is dataset with only instances with one or two cases

In [17]:
#df.to_csv('train1.csv', sep=',', encoding='utf-8', index=False)
#df_small.to_csv('train2.csv', sep=',', encoding='utf-8', index=False)